File: solutions/scala/meetup/1/src/main/scala/Meetup.scala
import LocalDateExtensions.LocalDateOps
import Schedule.{Last, Schedule, Teenth}

import java.time.{DayOfWeek, LocalDate}

case class Meetup(month: Int, year: Int) {
  private val TeenthOffset = 13
  private val DaysInWeek = 7
  private val FirstDayOfMonth = 1

  def day(targetWeekday: Int, schedule: Schedule): LocalDate =
    schedule match {
      case Teenth => findDay(targetWeekday, TeenthOffset, x => x)
      case Last   => findLastDay(targetWeekday)
      case _      => findDay(targetWeekday, FirstDayOfMonth, x => findByIndexedOrdinal(schedule, x))
    }

  private def dayOfWeekToInt(day: DayOfWeek): Int = day.getValue

  private def findLastDay(targetWeekday: Int): LocalDate = {
    val fourthTargetDay = day(targetWeekday, Schedule.Fourth)
    val candidateDate   = fourthTargetDay.plusWeek
    if (month == candidateDate.getMonthValue) {
      candidateDate
    } else {
      fourthTargetDay
    }
  }

  private def findByIndexedOrdinal(schedule: Schedule.Value, dayDifference: Int): Int =
    DaysInWeek * schedule.id + dayDifference

  private def findDay(targetWeekday: Int, baseDay: Int, baseOffset: Int => Int): LocalDate = {
    val nthOfMonth    = LocalDate.of(year, month, baseDay)
    val nthDayNumber  = dayOfWeekToInt(nthOfMonth.getDayOfWeek)
    val dayDifference = targetWeekday - nthDayNumber

    val candidateDay = nthOfMonth.plusDays(baseOffset(dayDifference))
    if (nthDayNumber <= targetWeekday) {
      candidateDay
    } else {
      candidateDay.plusWeek
    }
  }
}

object Meetup {
  val Mon: Int = DayOfWeek.MONDAY.getValue
  val Tue: Int = DayOfWeek.TUESDAY.getValue
  val Wed: Int = DayOfWeek.WEDNESDAY.getValue
  val Thu: Int = DayOfWeek.THURSDAY.getValue
  val Fri: Int = DayOfWeek.FRIDAY.getValue
  val Sat: Int = DayOfWeek.SATURDAY.getValue
  val Sun: Int = DayOfWeek.SUNDAY.getValue
}

File: solutions/erlang/meetup/1/src/meetup.erl
-module(meetup).

-export([meetup/4]).

-define(WEEKDAYS, [monday, tuesday, wednesday, thursday, friday, saturday, sunday]).
-define(ORDINAL, [first, second, third, fourth]).

meetup(Year, Month, TargetWeekday, teenth = _Schedule) ->
  find_day(
    Year,
    Month,
    TargetWeekday,
    _BaseDay = 13,
    fun(Offset) ->
      _BaseOffset = Offset
    end
  );

meetup(Year, TargetMonth, TargetWeekday, last = _Schedule) ->
  FourthTargetDay = meetup(Year, TargetMonth, TargetWeekday, fourth),
  CandidateDate = add_days_to_day(FourthTargetDay, 7),
  {_CandidateYear, CandidateMonth, _CandidateDay} = CandidateDate,

  if
    TargetMonth == CandidateMonth ->
      CandidateDate;
    true ->
      FourthTargetDay
  end;

meetup(Year, Month, TargetWeekday, Schedule) ->
  find_day(
    Year,
    Month,
    TargetWeekday,
    _BaseDay = 1,
    _BaseOffset = fun(Offset) ->
      find_by_indexed_ordinal(Schedule, Offset)
                  end
  ).

find_by_indexed_ordinal(Schedule, DayDifference) ->
  7 * (index_of(Schedule, ?ORDINAL) - 1) + DayDifference.

find_day(Year, Month, TargetWeekday, BaseDay, BaseOffsetFunction) ->
  {ok, NthOfMonth} = create_date(Year, Month, BaseDay),
  NthDayNumber = find_day_of_week(NthOfMonth),
  TargetDayNumber = to_day_number(TargetWeekday),
  DayDifference = TargetDayNumber - NthDayNumber,
  CandidateDay = add_days_to_day(NthOfMonth, BaseOffsetFunction(DayDifference)),

  if
    NthDayNumber =< TargetDayNumber ->
      CandidateDay;
    true ->
      add_days_to_day(CandidateDay, 7)
  end.

create_date(Year, Month, BaseDay) ->
  {ok, {Year, Month, BaseDay}}.

find_day_of_week(NthOfMonth) ->
  calendar:day_of_the_week(NthOfMonth).

add_days_to_day(Date, DaysToAdd) ->
  GregorianDays = calendar:date_to_gregorian_days(Date),
  calendar:gregorian_days_to_date(GregorianDays + DaysToAdd).

to_day_number(DayAtom) ->
  index_of(DayAtom, ?WEEKDAYS).

index_of(Item, List) -> index_of(Item, List, 1).

index_of(_, [], _) -> not_found;
index_of(Item, [Item | _], Index) -> Index;
index_of(Item, [_ | Tl], Index) -> index_of(Item, Tl, Index + 1).